packages/core/src/utils/treeTraversal.ts

```diff
@@ -0,0 +1,47 @@
+type NodeWithChildren<T> = {
+  children: T[];
+};
+
+export function treeVisit<T extends NodeWithChildren<T>>(
+  initialNode: T,
+  onNode: (node: T, index: number, depth: number) => void
+) {
+  // returns last used index
+  const _treeVisit = (currentNode: T, currentIndex: number, currentDepth: number): number => {
+    // Copy children in case of onNode removing it as we pass the node by reference
+    const children = [...currentNode.children];
+    onNode(currentNode, currentIndex, currentDepth);
+    let nextAvailableIndex = currentIndex + 1;
+    const lastUsedIndex = currentIndex;
+    for (const child of children) {
+      const lastUsedIndex = _treeVisit(child, nextAvailableIndex, currentDepth + 1);
+      nextAvailableIndex = lastUsedIndex + 1;
+    }
+    return lastUsedIndex;
+  };
+  _treeVisit(initialNode, 0, 0);
+}
+
+/**
+ * Traverses all nodes of a tree and maps all of them.
+ * Intuitively works similarly to Array.map().
+ * When returning mapped object from the mapper function
+ * you _do not_ need to return about `.children[]` field.
+ * This field is automatically managed by the treeMap() function.
+ *
+ * You can map each node to whatever you want, but shouldn't
+ * modify the structure of the tree from the mapping function.
+ */
+export function treeMap<T extends NodeWithChildren<T>, U>(
+  node: T,
+  onNode: (node: T) => Omit<U, 'children'>
+): U {
+  // Copy children in case of onNode removing it as we pass the node by reference
+  const children = [...node.children];
+  const newNode = {
+    ...onNode(node),
+    children: children.map((child) => treeMap(child, onNode)),
+  } as U;
+
```

Comment:
nit: i don't think we need to copy the children so this should also work like this

```
const { children, ...rest } = transformNode(node);
  return {
    ...rest,
    children: node.children.map(child => treeMap(child, transformNode)),
  } as U;
```

Reply:
Thanks for suggestion! 

I tried typing it in, but got some type problems (see the screenshot)

When writing this method original I recall it took quite a while to sort out types. And I even needed Thomas'es help to make types really fall in together. 

> Notice that `onNode()` isn't aware of `children`  field. 
![image](https://github.com/contentful/experience-builder/assets/113918657/0fdd8255-84eb-4bc7-bef1-2975c5b25ec0)


If you have ideas how to wrangle the types,  I put it in, but if types are hard to wrangle  I'd prefer leave it as it is working  🥺


![image](https://github.com/contentful/experience-builder/assets/113918657/88150554-80e5-4945-9d3e-9565318abf11)

 
 here's the snippet if you feel like you have time to play with it :)   
```ts
export function treeMap0<T extends NodeWithChildren<T>, U>(
  node: T,
  onNode: (node: T) => Omit<U, 'children'>
): U {
  // Copy children in case of onNode removing it as we pass the node by reference
  const { children, ...rest } = onNode(node);

  return {
    ...rest,
    children: node.children.map((child) => treeMap0(child, onNode)),
  };
}

```

Reply:
yeah no worries, that was just a nit so feel free to not do it if types are funky :)